### PR TITLE
Introduce CoresetMethodsSupport

### PR DESCRIPTION
**Pull request by francescodeaglio**
Many DeepCore methods require obtaining the last layer and using an embedding recorder. This PR introduces the CoresetMethodsSupport class that must be inherited in order to use these methods.

### CODE REVIEW

**Review by MaxiBoether — CHANGES_REQUESTED** [review]
Thanks! I think the biggest change is that we should have one ModynModel class which handles everything for Modyn models, not just recording embeddings. For example, the amp setting etc. This should be the first change and then I just had a few smaller comments. thanks again :)

File: modyn/models/modyn_model.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    def __exit__(self, *args: Any) -> None:
+        self.record_embedding = False
+        self.embedding = None
+
+
+class ModynModel(nn.Module, ABC):
+    def __init__(self, record_embedding: bool = False) -> None:
+        super().__init__()
+        self.embedding_recorder = EmbeddingRecorder(record_embedding)
```
[the remark this answers is not in the document]
Comment (francescodeaglio) [reply modyn/models/modyn_model.py:31]: Yes, because `record_embedding` can be set to `True` during execution (typically it is `False` during training and `True` when calculating scores)

File: modyn/models/modyn_model.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+# acknowledgment: github.com/PatrickZH/DeepCore/
+class EmbeddingRecorder(nn.Module):
+    def __init__(self, record_embedding: bool = False):
+        super().__init__()
+        self.record_embedding = record_embedding
+        self.embedding: Optional[torch.Tensor] = None
+
+    def forward(self, tensor: torch.Tensor) -> torch.Tensor:
+        if self.record_embedding:
+            self.embedding = tensor
```
[the remark this answers is not in the document]
Comment (francescodeaglio) [reply modyn/models/modyn_model.py:17]: In the original code they do not detach, I can investigate though

File: modyn/models/resnet18/resnet18.py [hunk 3]
```diff
@@ -1,10 +1,44 @@
...
         self.model.to(device)
+
+
+# the following class is adapted from
+# torchvision https://github.com/pytorch/vision/blob/main/torchvision/models/resnet.py
+
+
+class ResNetModyn(ResNet, ModynModel):
+    def __init__(self, model_configuration: dict[str, Any]) -> None:
+        super().__init__(BasicBlock, [2, 2, 2, 2], **model_configuration)
```
[the remark this answers is not in the document]
Comment (francescodeaglio) [reply modyn/models/resnet18/resnet18.py:22]: It's just the resnet we use (so resnet18)

File: modyn/models/resnet18/resnet18.py [hunk 4]
```diff
@@ -1,10 +1,44 @@
...
-from torchvision import models
+import torch
+from modyn.models.modyn_model import ModynModel
+from torch import Tensor, nn
+from torchvision.models.resnet import BasicBlock, ResNet
 
 
 class ResNet18:
     # pylint: disable-next=unused-argument
     def __init__(self, model_configuration: dict[str, Any], device: str, amp: bool) -> None:
```
[the remark this answers is not in the document]
Comment (francescodeaglio) [reply modyn/models/resnet18/resnet18.py:11]: I tried to generalise but every idea has problems. If I extend the functionality of `ModynModel` (e.g. device management) there are inheritance problems (e.g. `ResNet` inherits from both Torchvision's resnet and `ModynModel`) and it doesn't change much anyway (at most you can do model.to(device) but in more complex classes like `DLRM` it is almost useless as each part does it independently). If, on the other hand, ModynModel wraps the class (with something like `self.model = instantiate_class("modyn.models", class_name, ...)` you lose the purpose of `ModynModel` (i.e. to have abstract methods that have to be implemented in order to use certain methods for coresets). I think the best solution for now is to rename `ModynModel` to `CoresetMethodsSupport` and leave the refactor for a future PR

File: modyn/models/resnet18/resnet18.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (francescodeaglio) [reply modyn/models/resnet18/resnet18.py:22]: Now it's `ResNet18Modyn`

File: modyn/models/modyn_model.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    def __enter__(self) -> None:
+        self.record_embedding = True
+
+    def __exit__(self, *args: Any) -> None:
+        self.record_embedding = False
+        self.embedding = None
+
+
+class ModynModel(nn.Module, ABC):
```
[the remark this answers is not in the document]
Comment (francescodeaglio) [reply modyn/models/modyn_model.py:28]: Done

File: modyn/models/resnet18/resnet18.py [hunk 6]
```diff
@@ -1,10 +1,44 @@
...
+        x = self.avgpool(x)
+        x = torch.flatten(x, 1)
+        # the following line is the only difference compared to the original implementation
+        x = self.embedding_recorder(x)
+        x = self.fc(x)
+
+        return x
+
+    def get_last_layer(self) -> nn.Module:
+        return self.fc
```
[the remark this answers is not in the document]
Comment (francescodeaglio) [reply modyn/models/resnet18/resnet18.py:44]: To obtain information such as input and output size. It is taken from deepcore but may be removable

File: modyn/tests/models/test_dlrm.py [hunk 7]
```diff
@@ -67,3 +67,14 @@ def test_dlrm_reorder_categorical_input():
     assert reordered_test_data.shape == (64, 26)
     assert reordered_test_data.dtype == torch.long
     assert torch.equal(reordered_test_data, input_data)
+
+
+def test_get_last_layer():
```
[the remark this answers is not in the document]
Comment (francescodeaglio) [reply modyn/tests/models/test_dlrm.py:72]: added

File: modyn/models/modyn_model.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (francescodeaglio) [reply modyn/models/modyn_model.py]: No detaching since you need the computation graph to compute the gradient

File: modyn/models/resnet18/resnet18.py [hunk 8]
```diff
@@ -1,10 +1,44 @@
...
+        self.model = ResNetModyn(model_configuration)
         self.model.to(device)
+
+
+# the following class is adapted from
+# torchvision https://github.com/pytorch/vision/blob/main/torchvision/models/resnet.py
+
+
+class ResNetModyn(ResNet, ModynModel):
+    def __init__(self, model_configuration: dict[str, Any]) -> None:
```
[the remark this answers is not in the document]
Comment (francescodeaglio) [reply modyn/models/resnet18/resnet18.py:21]: Now it's more explicit since the context manager was difficult to use in the current codebase. You can call the method `model.embedding_recorder.start_recording()` to activate the recording and `model.embedding_recorder.end_recording()` to stop it. These methods will be called in the pytorch trainer during downsampling (`downsampling_batch` and `downsampling_trigger_training_set`) if the downsampling method requires `CoresetMethodsSupport`

File: modyn/selector/internal/selector_strategies/abstract_selection_strategy.py [hunk 9]
```diff
@@ -404,7 +404,7 @@ def get_available_labels(self) -> list[int]:
                 database.session.query(SelectorStateMetadata.label)
                 .filter(
                     SelectorStateMetadata.pipeline_id == self._pipeline_id,
-                    SelectorStateMetadata.seen_in_trigger_id >= self._next_trigger_id - self.tail_triggers
+                    SelectorStateMetadata.seen_in_trigger_id >= self._next_trigger_id - self.tail_triggers - 1
```
Comment (francescodeaglio) [inline modyn/selector/internal/selector_strategies/abstract_selection_strategy.py:407]: This query is done **after** the trigger (so `_next_trigger_id` has already been incremented) and thus we need the `-1` .  In other queries (for presampling strategies) there is no `-1` since they are performed before incrementing the counter

File: modyn/models/modyn_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/models/modyn_model.py]: ok

File: modyn/models/resnet18/resnet18.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/models/resnet18/resnet18.py:11]: Okay, I understand the problem. Then let's not have this within scope of this PR. I extended issue #48 with that

File: modyn/tests/models/test_dlrm.py [hunk 10]
```diff
@@ -67,3 +67,14 @@ def test_dlrm_reorder_categorical_input():
     assert reordered_test_data.shape == (64, 26)
```
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/tests/models/test_dlrm.py:67]: Thank you!

File: modyn/models/coreset_methods_support.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+    def start_recording(self) -> None:
+        self.record_embedding = True
+
+    def end_recording(self) -> None:
+        self.record_embedding = False
+        self.embedding = None
+
+
+class CoresetMethodsSupport(nn.Module, ABC):
```
Comment (MaxiBoether) [inline modyn/models/coreset_methods_support.py:28]: Maybe we should call this ´CoresetSupportingModule`?

File: modyn/models/resnet18/resnet18.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/models/resnet18/resnet18.py:21]: Sounds good and looks good! Thank you!

File: modyn/models/resnet18/resnet18.py [hunk 12]
```diff
@@ -1,10 +1,44 @@
...
-        self.model = models.__dict__["resnet18"](**model_configuration)
+        self.model = ResNet18Modyn(model_configuration)
         self.model.to(device)
+
+
+# the following class is adapted from
+# torchvision https://github.com/pytorch/vision/blob/main/torchvision/models/resnet.py
+
+
+class ResNet18Modyn(ResNet, CoresetMethodsSupport):
```
Comment (MaxiBoether) [inline modyn/models/resnet18/resnet18.py:20]: Any specific reason that Modyn is in the name here? I suppose we should make this consistent across the models we implement

File: integrationtests/selector/integrationtest_selector.py [hunk 13]
```diff
@@ -922,14 +922,15 @@ def test_get_available_labels(reset_after_trigger: bool):
         )
     )
 
-    selector.inform_data(
+    selector.inform_data_and_trigger(
```
Comment (MaxiBoether) [inline integrationtests/selector/integrationtest_selector.py:925]: I suppose this was an error and the change in the abstract selection strategy fixed this :)?

File: modyn/selector/internal/selector_strategies/abstract_selection_strategy.py (hunk 9 above)
Comment (MaxiBoether) [reply modyn/selector/internal/selector_strategies/abstract_selection_strategy.py:407]: just so I understand: `get_available_labels` should return all labels until the last trigger, correct? so if we have

add labels A,B,C -> trigger -> add label D -> get all labels

we should return A,B,C correct?

in the integration test you fixed, I see you added the trigger, but don't we also need to inform about a new data point (label D here) to ensure it is doing exactly that?

File: modyn/tests/selector/internal/selector_strategies/test_abstract_selection_strategy.py [hunk 14]
```diff
@@ -309,7 +309,7 @@ def test_get_available_labels_reset():
         database.session.commit()
 
     abstr = AbstractSelectionStrategy({"limit": -1, "reset_after_trigger": True}, get_minimal_modyn_config(), 1, 1000)
-
+    abstr._next_trigger_id += 1
```
Comment (MaxiBoether) [inline modyn/tests/selector/internal/selector_strategies/test_abstract_selection_strategy.py:312]: Could you add a comment here why this is necessary?

**Review by MaxiBoether — CHANGES_REQUESTED** [review]
Thanks! Almost done. I just had some confusion on the get_label thingy, after this is resolved we can merge this. Thank you!

File: modyn/models/resnet18/resnet18.py (hunk 12 above)
Comment (francescodeaglio) [reply modyn/models/resnet18/resnet18.py:20]: The usual name would be ResNet, but I cannot use it as I import it from Torch

File: modyn/tests/selector/internal/selector_strategies/test_abstract_selection_strategy.py (hunk 14 above)
Comment (francescodeaglio) [reply modyn/tests/selector/internal/selector_strategies/test_abstract_selection_strategy.py:312]: My initial idea was to simulate a trigger in this test. However, on re-reading it, I realised that it did not make much sense, so I simplified it by leaving the test just about getting the different labels.
The line was useless for this test.

File: integrationtests/selector/integrationtest_selector.py (hunk 13 above)
Comment (francescodeaglio) [reply integrationtests/selector/integrationtest_selector.py:925]: Exactly :)

Comment (francescodeaglio) [reply integrationtests/selector/integrationtest_selector.py:925]: And updated again :)

File: modyn/tests/selector/internal/selector_strategies/test_abstract_selection_strategy.py (hunk 14 above)
Comment (francescodeaglio) [reply modyn/tests/selector/internal/selector_strategies/test_abstract_selection_strategy.py:312]: Since the query relies on _next_trigger_id to filter out samples, it must be updated to simulate a real scenario (without triggers, the function always returns []). Incrementing the `_next_trigger_id` it's the only needed part of the trigger process.

File: modyn/selector/internal/selector_strategies/abstract_selection_strategy.py (hunk 9 above)
Comment (MaxiBoether) [reply modyn/selector/internal/selector_strategies/abstract_selection_strategy.py:409]: Thank you! This LGTM

**Review by MaxiBoether — APPROVED** [review]
Thank you, LGTM! Feel free to merge.
